cmd/replay.go

```diff
@@ -0,0 +1,208 @@
+	"github.com/pkg/errors"
+	cli "github.com/spf13/cobra"
+	"github.com/xlab/treeprint"
+	"google.golang.org/grpc"
+)
+
+type taskRunBlock struct {
+	name   string
+	height int
+	runs   set.Set
+}
+
+func taskRunBlockComperator(a, b interface{}) int {
+	aAsserted := a.(taskRunBlock)
+	bAsserted := b.(taskRunBlock)
+	switch {
+	case aAsserted.height < bAsserted.height:
+		return -1
+	case aAsserted.height > bAsserted.height:
+		return 1
+	}
+	return strings.Compare(aAsserted.name, bAsserted.name)
+}
+
+//formatRunsPerDAGInstance returns a hashmap with DAG -> Runs[] mapping
+func formatRunsPerDAGInstance(instance *pb.ReplayResponseNode, taskReruns map[string]taskRunBlock, height int) {
+	if _, ok := taskReruns[instance.JobName]; !ok {
+		taskReruns[instance.JobName] = taskRunBlock{
+			name:   instance.JobName,
+			height: height,
+			runs:   set.NewTreeSetWithTimeComparator(),
+		}
+	}
+
+	for _, taskRun := range instance.Runs {
+		parsedTime, _ := time.Parse(time.RFC3339, taskRun)
+		taskReruns[instance.JobName].runs.Add(parsedTime)
+	}
+	for _, child := range instance.Dependents {
+		formatRunsPerDAGInstance(child, taskReruns, height+1)
+	}
+}
+
+func replayCommand(l logger, conf *config.Optimus) *cli.Command {
+	cmd := &cli.Command{
+		Use:   "replay",
+		Short: "re-running jobs in order to update data for older dates/partitions",
+		Long:  `Backfill etl job and all of its downstream dependencies`,
+	}
+	cmd.AddCommand(replayRunSubCommand(l, conf))
+	return cmd
+}
+
+func replayRunSubCommand(l logger, conf *config.Optimus) *cli.Command {
+	dryRun := false
+	var (
+		replayProject string
+		namespace     string
+	)
+
+	reCmd := &cli.Command{
+		Use:     "run",
+		Short:   "run replay operation on a dag based on provided date range",
+		Example: "optimus replay run optimus.dag.name 2020-02-03 2020-02-05",
+		Long: `
+This operation takes three arguments, first is DAG name[required]
+used in optimus specification, second is start date[required] of
+replay, third is end date[optional] of replay. 
+Replay date ranges are inclusive.
+		`,
+		Args: func(cmd *cli.Command, args []string) error {
+			if len(args) < 1 {
+				return errors.New("dag name is required")
+			}
+			if len(args) < 2 {
+				return errors.New("replay start date is required")
+			}
+			return nil
+		},
+	}
+	reCmd.Flags().BoolVarP(&dryRun, "dry-run", "", dryRun, "do a trial run with no permanent changes")
+	reCmd.Flags().StringVarP(&replayProject, "project", "p", "", "project name of optimus managed ocean repository")
+	reCmd.MarkFlagRequired("project")
+	reCmd.Flags().StringVarP(&namespace, "namespace", "n", "", "namespace of deployee")
+	reCmd.MarkFlagRequired("namespace")
+
+	reCmd.Run = func(cmd *cli.Command, args []string) {
+		endDate := args[1]
+		if len(args) >= 3 {
+			endDate = args[2]
+		}
+		if err := replayJobBuildRequest(l, replayProject, namespace, args[0], args[1], endDate, dryRun, conf); err != nil {
+			l.Println(err)
+			os.Exit(1)
```

Comment:
No need to print error, no need to exit, just return error.

Reply:
updated